submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,354 @@
+    // If an empty namespaceSelector: {} is configured along with empty PodSelector, it applies
+    // to whole Cluster.
+    // On the other hand, if you omit specifying namespaceSelector, it does not select any
+    // namespaces and selects only Pods from the namespace where the GlobalnetEgressIP is
+    // deployed to.
+    // +optional
+    NamespaceSelector metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+    // Selects the pods whose label match the definition. This is an optional field and
+    // in case it is not set, it results in all the Pods selected from the NamespaceSelector.
+    // In case it is set, its intersected with NamespaceSelector, and only Pods that match the
+    // PodSelector and present in the NamespaceSelector will have the GlobalIPs as EgressIPs.
+    // Either namespaceSelector or podSelector have to be specified. If both are omitted,
+    // its considered as an error.
+    // When a Pod or Namespace matches multiple GlobalnetEgressIP objects, there is no guarantee
+    // which of the globalIP address that are assigned to GlobalnetEgressIP will be used.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+}
+
+const (
+    // GlobalnetEgressIPSuccess means that Globalnet was able to successfully
+    // allocate the GlobalIP as requested in GlobalnetEgressIP object.
+    GlobalnetEgressIPSuccess GlobalnetEgressIPStatus = "Success"
+
+    // GlobalnetEgressIPError means that Globalnet was unable to allocate the
+    // GlobalIP for the requested GlobalnetEgressIP object.
+    GlobalnetEgressIPError GlobalnetEgressIPStatus = "Error"
+)
+
+type EgressIPStatus struct {
+    // Status is one of {"Success", "Error"}
+    Status GlobalnetEgressIPStatus `json:"status,omitempty"`
+
+    // +optional
+    Message *string `json:"message,omitempty"`
+
+    // The list of GlobalIPs assigned via this GlobalnetEgressIP object.
+    GlobalIPs []string `json:"globalIPs"`
+
+    // The Namespaces to which the GlobalIPs are applied.
+    Namespaces []string `json:"namespaces"`
+
+    // The Pods to which the GlobalIPs are applied.
+    Pods []string `json:"pods"`
+}
+```
+
+When Submariner Globalnet is deployed in a Cluster, it auto-allocates the configured number of
+`globalIPs` and programs the necessary egress rules on the active Gateway node of the Cluster.
+This is done internally by creating GlobalnetEgressIP CR object with the name `cluster-default`.
+Users can query the CR to check the `Status.globalIPs` allocated at the Cluster level. If users
+create multiple GlobalnetEgressIP CRs that match the Cluster scope, one of the egressIP at the
+Cluster scope will be used.
+
+Creation of GlobalnetEgressIP object will be controlled via a ClusterRole and only users having
+the necessary role will be able to create the objects. The GlobalnetEgressIP object can be
+created in any namespace with selectors that match any other namespaces.
```

Comment:
@sridhargaddam 

Do we need to be clear who will have the ClusterRole?

I assume that there will be cluster admin and users for a particular namespace, at least.

Cluster admin obviously should be able to configure all of them.
On the other hand, users for a particular namespace isn't expected to be able to configure the one for other namespace in most cases.

Can we make it controllable or should we just don't allow latter use case? 
(Or do we allow users who have the ClusterRole to have control over all namespaces, as the current model does.)

Reply:
@mkimuram Does only admin will have access to create these objects?  unless a user is granted access for a specific reason,  going by the principle of least privilege.

Reply:
I think that some users will find it useful if they can configure IP addresses of the resources in their namespaces. And they basically won't care for other namespaces. If we can allow them to configure it only for their namespace, it should satisfy the principle of least privilege. 

If we only allow admin to configure it, which is not what I expect, we should consider making this CRD cluster scoped resource or allowing it to exist only in a specific namespace, like submariner-operator namespace. It would make it easier to check what is configured.  

Then, should we consider separating this CRD to one for admin to configure cluster level and the other for user to configure namespace level?

Reply:
We discussed this point in today's external network meeting and agreed on the following. @mkolesnik/@aswinsuryan  if you have any comments on this, do let us know.

The existing `GlobalnetEgressIP` CRD will be modified to Cluster scope and only admin will have the provision to create the objects that apply to various namespaces. In future, based on the feedback, we can introduce a different CRD which does NOT include the NamespaceSelectors and can be applied at namespace level. Such objects can be created by regular users based on the clusterRoles. This will be less confusing IMHO.

Reply:
Did you discuss https://github.com/submariner-io/enhancements/pull/22/files#r587478495 ?

IMHO it's much more intuitive and a better UX then what is proposed, BUT if it's already been discussed and rejected then we could go with the agreed upon solution (cluster level CRs).

I think the main point here is the UX and RBAC - should a user who's not an admin be able to define a globalnet policy for his namespace/pods?
It seems reasonable to me not to have to rely on admin role for this, as the cloud & k8s models are much about self service and not relying on request fulfilment from 3rd party (ie some admin or another).

To simplify we could have cluster level policy not set by this CRD at all since it's really a singleton policy for the cluster, and the CRD could be namespace scoped and define a policy for pods in that namespace.

On the UX I imagine this would be abstracted from the user anyway by subctl, so let's get the right design that makes most sense.

Reply:
> It seems reasonable to me not to have to rely on admin role for this, as the cloud & k8s models are much about self service and not relying on request fulfilment from 3rd party (ie some admin or another).

Agree, it makes sense and aligns more with K8s models. 

```
Perhaps a more coherent model would be that a CR on a namespace would only affect pods in that namespace, and if no pod selector is specified then it affects all pods in the namespace?
Then namespace selector is implicit and doesn't need to exist as a field.

Cluster wide settings can then be applied on the submariner namespace in a similar fashion allowing either all pods or select pods.
```
Taking a re-look at your suggestion, if I understood it correctly, you are suggesting that if admin creates the object in `submariner-operator` namespace then we treat it as `cluster-wide`. Is this correct @mkolesnik ?

Assuming yes, what I see missing is that we do not have provision to create objects that select specific `namespaces` (as we will be removing `Spec.NamespaceSelector`). Personally, this looks okay to me and we can add this support in a later release if required. CC: @mangelajo @mkimuram @vthapar @aswinsuryan 
 
> To simplify we could have cluster level policy not set by this CRD at all since it's really a singleton policy for the cluster, and the CRD could be namespace scoped and define a policy for pods in that namespace.

I did not understand this completely. Will try to sync up with you over slack.

Reply:
I discussed this with @mkolesnik and @vthapar over slack and we felt that it's better to create a separate CR that is cluster-wide and name it as `ClusterGlobalEgressIP`. 

At the same time rename the current CR to `NamespaceGlobalEgressIP` 

Only admins will have the provision to create the `ClusterGlobalEgressIP` CR and regular users will only be able to create `NamespaceGlobalEgressIP` that applies to their own namespaces.